mmcv/runner/ipu_runner.py

```diff
@@ -0,0 +1,136 @@
+        ipu_model_cfg = ipu_model_cfg or {}
+        super(IPUBaseRunner, self).__init__(model, **kwargs)
+
+        # process options of ipu
+        if IPU_MODE:
+            self.ipu_options = parse_ipu_options(ipu_options)
+            # self.data_loader = wrap_data_loader(self.data_loader)
+            self.model = ipu_model_wrapper(
+                self.model, self.ipu_options, self.optimizer, self.logger,
+                modules_to_record=modules_to_record,
+                ipu_model_cfg=ipu_model_cfg, fp16_cfg=fp16_cfg)
+        else:
+            # warnings.warn('no ipu found, degrade to CPU mode', UserWarning)
+            raise NotImplementedError('cpu mode on IPURunner not supported')
+
+    def register_lr_hook(self, lr_config):
+        assert isinstance(lr_config, dict)
+        assert 'policy' in lr_config
+        policy_type = lr_config.pop('policy')
+        # If the type of policy is all in lower case,
+        # e.g., 'cyclic', then its first letter will be capitalized,
+        # e.g., to be 'Cyclic'.
+        # This is for the convenient usage of Lr updater.
+        # Since this is not applicable for `
+        # CosineAnnealingLrUpdater`, the string will not be changed
+        # if it contains capital letters.
+        if policy_type == policy_type.lower():
+            policy_type = policy_type.title()
+        hook_type = policy_type + 'LrUpdaterHook'
+        lr_config['type'] = hook_type
+        hook = build_from_cfg_with_wrapper(
+            lr_config, HOOKS, wrap_lr_update_hook)
+        self.register_hook(hook, priority='VERY_HIGH')
+
+    def register_optimizer_hook(self, optimizer_config):
+        assert isinstance(optimizer_config, (dict, IPUFp16OptimizerHook))
+        if isinstance(optimizer_config, dict):
+            optimizer_config.setdefault('type', 'OptimizerHook')
+            hook = build_from_cfg_with_wrapper(
+                optimizer_config, HOOKS, wrap_optimizer_hook)
+        else:
+            hook = optimizer_config
+        self.register_hook(hook, priority='ABOVE_NORMAL')
+
+    def run(self, data_loaders, workflow, *args, **kwargs):
+        for i, flow in enumerate(workflow):
+            mode, iters = flow
+            # initialize IPU dataloder if not initialized
+            data_loaders[i].init(options=self.get_ipu_opts(mode))
+
+        super().run(data_loaders, workflow, *args, **kwargs)
+
+    def get_ipu_opts(self, mode):
+        if mode == 'train':
+            return self.ipu_options['training']
+        elif mode == 'val':
+            return self.ipu_options['inference']
+
+
+@RUNNERS.register_module()
+class IPUEpochBasedRunner(IPUBaseRunner, EpochBasedRunner):
+    """Epoch-based Runner for IPU.
+    The Inheritance order(MRO) is:
+    IPUEpochBasedRunner -> IPUBaseRunner -> EpochBasedRunner -> BaseRunner
+    This runner train models epoch by epoch.
+    """
+    pass
+
+
+@RUNNERS.register_module()
+class IPUIterBasedRunner(IPUBaseRunner, IterBasedRunner):
```

Comment:
Can we simplify this inheritance with `single inheritance`?

Reply:
no idea how to simplify.
because we need inheritance from IterBasedRunner and IPUBaseRunner.
or we need to implement same functions twice in IPUIterBasedRunner and IPUEpochBasedRunner.